## Sort centrality measures and determine most influential web developers and machine learning engineers

In [7]:
import pandas as pd

df = pd.read_csv('Github_degree_betweenness_closeness.csv')

df['ml_target'].value_counts()
df['Degree_betweenness_closeness'] = 1/3*(df['Degree'] + df['betweenness'] + df['closeness'])

In [8]:
df_ml = df[df['ml_target']==1]
df_wd = df[df['ml_target']==0]



In [9]:
df_ml.sort_values(by = 'Degree_betweenness_closeness', ascending = False, inplace=True)
df_ml.head()

C:\Users\Thomas\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Account_name    Degree  betweenness  closeness     id        name  \
65           23589  0.025651     0.007242   0.401758  23589     antirez   
184          22321  0.016393     0.003188   0.402492  22321    bradfitz   
3457         26649  0.007295     0.002157   0.404908  26649   WillemJan   
1366         14954  0.019337     0.008232   0.378599  14954       rasbt   
877           1089  0.005119     0.000771   0.400087   1089  iam-peekay   

      ml_target  Degree_betweenness_closeness  
65            1                      0.144884  
184           1                      0.140691  
3457          1                      0.138120  
1366          1                      0.135389  
877           1                      0.135326

In [10]:
df_wd.sort_values(by = 'Degree_betweenness_closeness', ascending = False, inplace = True)
df_wd.head()

C:\Users\Thomas\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Account_name    Degree  betweenness  closeness     id               name  \
28          31890  0.250882     0.269599   0.517787  31890       dalinhuang99   
24          27803  0.187936     0.240541   0.523081  27803             nfultz   
82          19222  0.078464     0.055323   0.466324  19222            Bunlong   
35          35773  0.088172     0.043408   0.450342  35773         addyosmani   
109         13638  0.065466     0.035337   0.447461  13638  gabrielpconceicao   

     ml_target  Degree_betweenness_closeness  
28           0                      0.346089  
24           0                      0.317186  
82           0                      0.200037  
35           0                      0.193974  
109          0                      0.182755

## Determine homophily among the network

In [12]:
df_homophily = pd.read_csv('combined_github_data.csv')
df_homophily.head()


Unnamed: 0  id_1   id_2        name  ml_target       name_2  ml_target_2
0           0     0  23977      Eiryyy          0     airtoxin            0
1           1     1  34526  shawflying          0      ghosind            0
2           2     1   2370  shawflying          0      jasondu            0
3           3     1  14683  shawflying          0  chaoslawful            0
4           4     1  29982  shawflying          0   dead-horse            0

In [19]:
set1 = set(df_homophily['name'])
set2 = set(df_homophily['name_2'])
nodes = set1.union(set2)
len(nodes)

37700

In [20]:
df_accounts = pd.read_csv('musae_git_target.csv')
len(df_accounts)

37700

In [29]:
prob_ml = sum(df_accounts['ml_target'])/len(df_accounts)
prob_wd = 1-prob_ml
prob_cross_edge = 2*prob_ml*prob_wd

print('The expected probability of a cross-edge is:')
print(prob_cross_edge)

mask = df_homophily['ml_target'] != df_homophily['ml_target_2']
df_cross_edges = df_homophily[mask]
actual_prob = len(df_cross_edges) / len(df_homophily)
print('The actual probability of a cross-edge is:')
print(actual_prob)

The expected probability of a cross-edge is:
0.38319017090108276
The actual probability of a cross-edge is:
0.1546558340224842


## Since the actual probability of a cross-edge (0.15) is less than the expected probability of a cross edge (0.38), there is homophily in the network (i.e., MLEs tend to work together more than with web developers and vice versa)